relay-server/src/envelope.rs

```diff
@@ -1080,6 +1080,38 @@ impl Envelope {
         }))
     }
 
+    /// Splits the envelope by the given predicate.
+    ///
+    /// The main differents from `split_by()` is this function returns the list of the newly
+    /// constracted envelopes with all the items where the predicate returns `true`. Otherwise it
+    /// returns an empty list.
+    ///
+    /// The returned envelopes assume the same headers.
+    pub fn split_all_by<F>(&mut self, mut f: F) -> SmallVec<[Box<Self>; 3]>
+    where
+        F: FnMut(&Item) -> bool,
+    {
+        let mut envelopes = smallvec![];
+
+        let split_count = self.items().filter(|item| f(item)).count();
+        if split_count == 0 {
+            return envelopes;
+        }
+
+        let old_items = std::mem::take(&mut self.items);
+        let (split_items, own_items) = old_items.into_iter().partition(f);
+        self.items = own_items;
+
+        for item in split_items {
+            envelopes.push(Box::new(Envelope {
+                headers: self.headers.clone(),
+                items: smallvec![item],
+            }))
+        }
+
+        envelopes
```

Comment:
nit: `split_by` and `split_all_by` could use a common helper method `split_off_items` that returns `split_items`.

Reply:
Will look into it